docs/source/sharing_intermediates.rst

```diff
@@ -0,0 +1,36 @@
+=====================
+Sharing Intermediates
+=====================
+
+If you want to compute multiple similar contractions with common terms, you can embed them in a :func:`~opt_einsum.shared_intermediates` context. Computations of subexpressions in this context will be memoized, and will be garbage collected when the contexts exits.
+
+For example, suppose we want to compute marginals at each point in a factor chain:
+
+.. code:: python
+
+    >>> inputs = 'ab,bc,cd,de,ef'
+    >>> factors = [np.random.rand(4, 4) for _ in range(5)]
+    >>> marginals = {output: contract('{}->{}'.format(inputs, output), *factors)
+    >>>              for output in 'abcdef'}
+
+To share this computation, we can perform all contractions in a shared context
+
+.. code:: python
+
+    >>> with shared_intermediates():
+    >>>     marginals = {output: contract('{}->{}'.format(inputs, output), *factors)
+    >>>                  for output in 'abcdef'}
```

Comment:
Might be worth showing a quick timing comparison on bigger tensors with an explicit demonstration of the contractions done.